# **Data Cleaning sample and Working with Local Database connection**

In [1]:
# Import modules
import sys
import os
import pandas as pd

In [2]:
# Add the 'scripts' directory to the Python path for module imports
sys.path.append(os.path.abspath(os.path.join('..', 'scripts')))

In [3]:
# Now import the modules
from data_cleaning import load_csv, clean_dataframe, save_cleaned_data

### **Load and Inspect Raw Data**

In [4]:
df = load_csv("../data/messages/scraped_messages.csv")

# Show first few rows
df.head(10)


channel_name     channel_title                 date  \
0    DoctorsET  Doctors Ethiopia  2023-12-18 17:04:02   
1    DoctorsET  Doctors Ethiopia  2023-11-03 16:14:39   
2    DoctorsET  Doctors Ethiopia  2023-10-02 16:37:39   
3    DoctorsET  Doctors Ethiopia  2023-09-16 07:54:32   
4    DoctorsET  Doctors Ethiopia  2023-09-01 16:16:15   
5    DoctorsET  Doctors Ethiopia  2023-08-29 17:20:05   
6    DoctorsET  Doctors Ethiopia  2022-08-02 17:42:08   
7    DoctorsET  Doctors Ethiopia  2022-06-12 17:15:47   
8    DoctorsET  Doctors Ethiopia  2022-05-31 17:51:13   
9    DoctorsET  Doctors Ethiopia  2022-05-20 18:04:53   

                                                text  
0  https://youtu.be/5DBoEm-8kmA?si=LDLuEecNfULJVD...  
1  ዶክተርስ ኢትዮጵያ በ አዲስ አቀራረብ በ ቴሌቪዥን ፕሮግራሙን ለመጀመር ከ...  
2  ሞት በስኳር \n\nለልጆቻችን የምናሲዘው ምሳቃ ሳናቀው እድሚያቸውን ይቀን...  
3  ከ HIV የተፈወሰ ሰው አጋጥሟችሁ ያቃል ? ፈውስ እና ህክምና ?\n\nሙ...  
4  በቅርብ ጊዜ በሃገራችን ላይ እየተስተዋለ ያለ የተመሳሳይ ፆታ ( Homos...  
5  👇👇👇👇👇👇 https://youtu.be/-AR1KO2DbFw?si=47cXLZt...  
6  ክረምቱን ስፖርት መስራት አስበው ጂም ለመግባት ካልቻሉ ባሉበት ቦታ ሆነው...  
7  ስፖርት የመስራት ሱስ ይኖር ይሆን?\n\nበአሁኑ ወቅት ብዙ የስፖርት መስ...  
8  ድንገተኛ አደጋ / የአጥንት ስብራት\n\nአያርገውና ድንገተኛ የሆነ አደጋ...  
9  ከትንሽ ግዚያት በፊት ስፖርት መስራት እንደ ቅንጦት ይታይ ነበር አሁን ላ...

### **Clean and Standardize Data**

In [5]:
df_cleaned = clean_dataframe(df)

# Display cleaned dataset
df_cleaned.head(50)

channel_name     channel_title            text_date  \
0     DoctorsET  Doctors Ethiopia  2023-12-18 17:04:02   
1     DoctorsET  Doctors Ethiopia  2023-11-03 16:14:39   
2     DoctorsET  Doctors Ethiopia  2023-10-02 16:37:39   
3     DoctorsET  Doctors Ethiopia  2023-09-16 07:54:32   
4     DoctorsET  Doctors Ethiopia  2023-09-01 16:16:15   
5     DoctorsET  Doctors Ethiopia  2023-08-29 17:20:05   
6     DoctorsET  Doctors Ethiopia  2022-08-02 17:42:08   
7     DoctorsET  Doctors Ethiopia  2022-06-12 17:15:47   
8     DoctorsET  Doctors Ethiopia  2022-05-31 17:51:13   
9     DoctorsET  Doctors Ethiopia  2022-05-20 18:04:53   
10    DoctorsET  Doctors Ethiopia  2022-05-15 15:59:10   
11    DoctorsET  Doctors Ethiopia  2022-05-07 18:22:14   
12    DoctorsET  Doctors Ethiopia  2022-05-06 17:51:05   
13    DoctorsET  Doctors Ethiopia  2022-05-06 17:51:05   
14    DoctorsET  Doctors Ethiopia  2022-05-06 17:51:05   
15    DoctorsET  Doctors Ethiopia  2022-05-06 17:51:05   
16    DoctorsET  Doctors Ethiopia  2022-05-01 11:13:33   
17    DoctorsET  Doctors Ethiopia  2022-04-29 18:59:05   
18    DoctorsET  Doctors Ethiopia  2022-04-29 18:57:35   
19    DoctorsET  Doctors Ethiopia  2022-04-28 05:46:38   
20    DoctorsET  Doctors Ethiopia  2022-04-28 05:46:38   
21    DoctorsET  Doctors Ethiopia  2022-04-28 05:46:38   
22    DoctorsET  Doctors Ethiopia  2022-04-28 05:46:38   
23    DoctorsET  Doctors Ethiopia  2022-04-24 17:11:09   
24    DoctorsET  Doctors Ethiopia  2022-04-22 18:55:31   
25    DoctorsET  Doctors Ethiopia  2022-04-22 17:51:05   
26    DoctorsET  Doctors Ethiopia  2022-04-22 17:51:05   
27    DoctorsET  Doctors Ethiopia  2022-04-22 17:51:04   
28    DoctorsET  Doctors Ethiopia  2022-04-22 17:51:04   
29    DoctorsET  Doctors Ethiopia  2022-04-22 17:51:04   
30    DoctorsET  Doctors Ethiopia  2022-04-22 17:51:04   
31    DoctorsET  Doctors Ethiopia  2022-04-19 17:38:05   
32    DoctorsET  Doctors Ethiopia  2022-04-15 18:06:19   
33    DoctorsET  Doctors Ethiopia  2022-04-15 18:06:19   
34    DoctorsET  Doctors Ethiopia  2022-04-15 18:06:19   
35    DoctorsET  Doctors Ethiopia  2022-04-14 04:14:11   
36    DoctorsET  Doctors Ethiopia  2022-04-12 18:01:18   
37    DoctorsET  Doctors Ethiopia  2022-04-08 17:07:21   
38    DoctorsET  Doctors Ethiopia  2022-04-08 17:07:21   
39    DoctorsET  Doctors Ethiopia  2022-04-08 17:07:21   
40    DoctorsET  Doctors Ethiopia  2022-04-08 17:07:21   
41    DoctorsET  Doctors Ethiopia  2022-04-07 04:24:11   
42    DoctorsET  Doctors Ethiopia  2022-04-07 04:24:11   
43    DoctorsET  Doctors Ethiopia  2022-04-07 04:24:11   
44    DoctorsET  Doctors Ethiopia  2022-04-07 04:24:11   
45    DoctorsET  Doctors Ethiopia  2022-04-03 16:15:29   
46    DoctorsET  Doctors Ethiopia  2022-04-02 06:06:36   
47    DoctorsET  Doctors Ethiopia  2022-04-01 17:02:54   
48    DoctorsET  Doctors Ethiopia  2022-04-01 17:02:54   
49    DoctorsET  Doctors Ethiopia  2022-04-01 17:02:54   

                                              message      emoji_used  
0   በቀን አንዴ ብቻ የሚባለው የቢዝነስ አማካሪ  በ 10,000 ብር ብቻ የተ...            👈👈👇👇  
1   ዶክተርስ ኢትዮጵያ በ አዲስ አቀራረብ በ ቴሌቪዥን ፕሮግራሙን ለመጀመር ከ...               👇  
2   ሞት በስኳር  ለልጆቻችን የምናሲዘው ምሳቃ ሳናቀው እድሚያቸውን ይቀንሰው ...        No emoji  
3   ከ HIV የተፈወሰ ሰው አጋጥሟችሁ ያቃል ? ፈውስ እና ህክምና ? ሙሉ ቪ...            👇👇👇👇  
4   በቅርብ ጊዜ በሃገራችን ላይ እየተስተዋለ ያለ የተመሳሳይ ፆታ ( Homos...        No emoji  
5   ** ዶክተርስ ኢትዮጽያ በአዲስ ፕሮገራም ጀመረ** ማረጥ  (**ሜኖፖዝ )...          👇👇👇👇👇👇  
6   ክረምቱን ስፖርት መስራት አስበው ጂም ለመግባት ካልቻሉ ባሉበት ቦታ ሆነው...           👇👇👇👇👇  
7   ስፖርት የመስራት ሱስ ይኖር ይሆን? በአሁኑ ወቅት ብዙ የስፖርት መስሪያ ...          👇👇👇👇👇👇  
8   ድንገተኛ አደጋ / የአጥንት ስብራት አያርገውና ድንገተኛ የሆነ አደጋ ቢደ...         👇👇👇👇👇👇👇  
9   ከትንሽ ግዚያት በፊት ስፖርት መስራት እንደ ቅንጦት ይታይ ነበር አሁን ላ...          👇👇👇👇👇👇  
10  **ልጆች ላላችሁ / አባት እናት ወይንም እህት ወንድም / በቅርቡ ልጅ ለ...            👇👇👇👇  
11  **በብዙ ማህበራዊ ሚዲያዎች በ ከፋተኛ ሁኔታ እየተሸጠ ያለው  ፕሮቲን- ...  👇👇👇👇👇👇👇👇👇👇👇👇👇👇  
12                                          [No Text]        No emoji  
13                                  

In [6]:
# Check for missing values in the cleaned DataFrame
missing_values = df_cleaned.isnull().sum()
missing_values[missing_values > 0] 

Series([], dtype: int64)

In [7]:
# Get the dimensions of the DataFrame: (number of rows, number of columns)
df_cleaned.shape

(3828, 5)

The output (3828, 5) indicates that the DataFrame contains 3828 rows and 5 columns.

In [8]:
# Get the total number of elements in the DataFrame (rows * columns)
df_cleaned.size

19140

### **Save the Cleaned Data**

In [9]:
# Save cleaned data to CSV
save_cleaned_data(df_cleaned, "../data/processed/cleaned_telegram_data.csv")

✅ Cleaned data saved successfully to '../data/processed/cleaned_telegram_data.csv'.


# **Connect to Database**

In [13]:
from database_setup import get_db_connection, create_table, insert_data
engine = get_db_connection()

###  **Create Table in PostgreSQL**

In [14]:
create_table(engine)

###  **Insert Data into Database**

In [15]:
# Load the cleaned CSV into a DataFrame
cleaned_df = pd.read_csv("../data/cleaned_telegram_data.csv")

In [16]:
# Ensure the 'message_date' column is in datetime format (to prevent NaT issues)
cleaned_df["text_date"] = pd.to_datetime(cleaned_df["text_date"], errors="coerce")

# Check if there are any missing values before inserting
missing_values = cleaned_df.isnull().sum()
print("Missing Values Before Insert:", missing_values)

Missing Values Before Insert: channel_name     0
channel_title    0
text_date        0
message          3
emoji_used       0
dtype: int64
